### ISSUE TITLE
given should not accept any special keyword arguments

### ISSUE THREAD

**Issue by DRMacIver** [HypothesisWorks/hypothesis-python#180]
Current status: given has two special keyword arguments, settings and random. Both of these are basically a recipe for tripping you up if you wanted to have arguments of those names and will cause a bunch of confusion when that happens. It also means we can never add additional arguments - something which would be very useful for e.g. #163 

These arguments should be deprecated and a new API for passing them provided. I don't currently have a strong opinion on what that should look like but perhaps something like:

```python
configuration = Configuration(settings=Settings, random=Random)

@configuration.given(integers())
def test_foo_bar(x):
   ....
```

Configuration as a distinct concept from settings isn't great, but things like Random or a lifecycle callback really don't belong in settings.

I'm currently not thrilled by any of the options which is why this hasn't been done yet. Suggestions very welcome.

**Comment by DRMacIver** [comment]
Suggestion from @The-Compiler on IRC which I think is now my favourite solution:

```python
@given(integers())
@configure(settings=Settings(), random=Random)
def test_foo_bar(x):
    pass
```

As with ```@example``` by storing an attribute that is used when the function is run, this can be put either before or after the ```@given``` as you prefer.

**Comment by kxepal** [comment]
```
@given(integers())
@with_settings(Settings, Random)
def test_foo_bar(x):
    ...
```

And this actually helps to keep keyword arguments with no conflicts:
```
@given(x=integers())
@with_settings(Settings, Random)
def test_foo_bar(x):
    ...
```

Why not `with Settings():` ? Because if one day I decide to remove it, I have to reindent the whole block which will not make changes history clearer.

**Comment by DRMacIver** [comment]
Snap. :-) And yeah, I agree that using context managers here isn't a great idea.

**Comment by DRMacIver** [comment]
This has been fixed for a while (since 2.0)
